From 8799a1d2816f33108f737da4aa2014cfc16dda91 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Tue, 18 Sep 2018 09:32:23 +0300
Subject: [PATCH] preboot: hard code the SPL size

Currently, the SPL size is calculated during the build of U-Boot/SPL
and is passed to BTL build. The BTL build in turn, saves the SPL size
in its code and writes it into the scratchpad 29 register.
Preboot reads the SPL size from the scratchpad 29 and uses it to load
SPL into SRAM.
The above creates a dependency between SPL and BTL builds, and in
addition is error prone during the changes between the primary and the
secondary during update process.
To make the update process more robust and remove the build dependency,
use the hard coded value of 64KB for loading the SPL.
It is a size of an SPI flash sector. It is big enough to cope with SPL
growth (if any) and it is small enough to not impact the performance
during boot.

Change-Id: I12b16fe4cb9359d56d6486a635cf861035a3f44a
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/preboot.c b/preboot.c
index c5cc07b..ee6cb21 100644
--- a/preboot.c
+++ b/preboot.c
@@ -8,11 +8,13 @@
 #include <config.h>
 #include <common.h>
 #include <linux/io.h>
+#include <linux/sizes.h>
 #include "mach/hl_regs.h"
 
 #define SYS_REF_CLOCK			50000000
 #define BTL_COUNTER_MAX			BIT(22)
 #define mmPSOC_UART_0_DATA		(HL_CFG_BASE + 0xC45000)
+#define SPL_SIZE			SZ_64K
 
 /* Use DumpSel[8] to skip PCIe initialization */
 #ifdef DEV_IS_GOYA
@@ -67,14 +69,13 @@
 static void hl_preboot_copy_spl(void)
 {
 	void *spl_src, *spl_dst;
-	unsigned int i, spl_offset, spl_size;
+	unsigned int i, spl_offset;
 
 	hl_preboot_copy_prepare(&spl_offset);
 	spl_src = (void*)HL_SPI_BASE + spl_offset;
-	spl_size = readl(mmPSOC_GLOBAL_CONF_SCRATCHPAD_29);
 	spl_dst = (void*)CONFIG_SPL_TEXT_BASE;
 
-	for (i=0; i < spl_size; i += sizeof(unsigned long))
+	for (i=0; i < SPL_SIZE; i += sizeof(unsigned long))
 		*(unsigned long*)(spl_dst + i) = *(unsigned long*)(spl_src + i);
 }
 
